modules/node-shared/src/main/scala/io/constellationnetwork/node/shared/domain/delegatedStake/UpdateDelegatedStakeValidator.scala

```diff
@@ -0,0 +1,198 @@
+import derevo.derive
+import eu.timepit.refined.types.numeric.NonNegLong
+
+trait UpdateDelegatedStakeValidator[F[_]] {
+  def validateCreateDelegatedStake(
+    signed: Signed[UpdateDelegatedStake.Create],
+    lastContext: GlobalSnapshotInfo,
+    lastGlobalEpochProgress: EpochProgress
+  ): F[UpdateDelegatedStakeValidationErrorOr[Signed[UpdateDelegatedStake.Create]]]
+  def validateWithdrawDelegatedStake(
+    signed: Signed[UpdateDelegatedStake.Withdraw],
+    lastContext: GlobalSnapshotInfo
+  ): F[UpdateDelegatedStakeValidationErrorOr[Signed[UpdateDelegatedStake.Withdraw]]]
+}
+
+object UpdateDelegatedStakeValidator {
+  def make[F[_]: Async: SecurityProvider](
+    signedValidator: SignedValidator[F],
+    seedlist: Option[Set[SeedlistEntry]],
+    minEpochProgressesToLock: NonNegLong
+  )(implicit hasher: Hasher[F]): UpdateDelegatedStakeValidator[F] =
+    new UpdateDelegatedStakeValidator[F] {
+
+      def validateCreateDelegatedStake(
+        signed: Signed[UpdateDelegatedStake.Create],
+        lastContext: GlobalSnapshotInfo,
+        lastGlobalEpochProgress: EpochProgress
+      ): F[UpdateDelegatedStakeValidationErrorOr[Signed[UpdateDelegatedStake.Create]]] =
+        for {
+          numberOfSignaturesV <- validateNumberOfSignatures(signed)
+          signaturesV <- signedValidator
+            .validateSignatures(signed)
+            .map(_.errorMap[UpdateDelegatedStakeValidationError](InvalidSigned))
+          nodeIdV = validateNodeId(signed)
+          tokenLockV <- validateTokenLock(signed, lastContext, lastGlobalEpochProgress)
+        } yield numberOfSignaturesV.productR(signaturesV).productR(nodeIdV).productR(tokenLockV)
+
+      def validateWithdrawDelegatedStake(
+        signed: Signed[UpdateDelegatedStake.Withdraw],
+        lastContext: GlobalSnapshotInfo
+      ): F[UpdateDelegatedStakeValidationErrorOr[Signed[UpdateDelegatedStake.Withdraw]]] =
+        for {
+          numberOfSignaturesV <- validateNumberOfSignatures(signed)
+          signaturesV <- signedValidator
+            .validateSignatures(signed)
+            .map(_.errorMap[UpdateDelegatedStakeValidationError](InvalidSigned))
+          nodeIdV = validateNodeId(signed)
+          withdrawV <- validateWithdrawal(signed, lastContext)
+        } yield numberOfSignaturesV.productR(signaturesV).productR(nodeIdV).productR(withdrawV)
+
+      private def validateNumberOfSignatures[A <: UpdateDelegatedStake](
+        signed: Signed[A]
+      ): F[UpdateDelegatedStakeValidationErrorOr[Signed[A]]] = {
+        val result = if (signed.proofs.size == 1) {
+          signed.validNec[UpdateDelegatedStakeValidationError]
+        } else {
+          TooManySignatures(signed.proofs).invalidNec
+        }
+        result.pure[F]
+      }
+
+      private def validateNodeId[A <: UpdateDelegatedStake](
+        signed: Signed[A]
+      ): UpdateDelegatedStakeValidationErrorOr[Signed[A]] =
+        if (seedlist.forall(_.exists(_.peerId == signed.nodeId))) {
```

Comment:
Nitpick comment: I realize elsewhere in the code we're following this pattern as treating seedlist as an Option type and enforcing it elsewhere, but it would be much better if the tests mocked up a seedlist based on the generators, rather than allowing this to be optional everywhere, since it's essentially always required.

It's not required for you to change this, but if you can see a place to introduce a pattern it would be helpful.